### PR TITLE
VIH-9013 - Refactor Terraform Pipelines

### PR DESCRIPTION
**Pull request by v1ferrarij**
### JIRA link (if applicable) ###

[VIH-9013](https://tools.hmcts.net/jira/browse/VIH-9013)

### Change description ###

Refactor Terraform Pipelines for Infra Core & Wowza.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X ] No
```

### CODE REVIEW

File: azure-pipelines.infra-core.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+   - Dev
+   - Demo
+
+pool:
+  vmImage: 'ubuntu-latest'
+
+variables:
+- name: tfVersion
+  value: 1.1.0
+- name: tfStateServiceConnection
```
Comment (chris-pateman) [inline azure-pipelines.infra-core.yaml:25]: Do we need this if state is held in each environment?

File: azure-pipelines.infra-core.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
+resources:
+  repositories:
+    - repository: templates
+      type: github
+      name: hmcts/azure-devops-templates
+      ref: refs/heads/master
+      endpoint: hmcts
+
+trigger:
```
Comment (chris-pateman) [inline azure-pipelines.infra-core.yaml:9]: Do we want merged into master trigger and pr trigger?

File: azure-pipelines.infra-core.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+- name: tfVersion
+  value: 1.1.0
+- name: tfStateServiceConnection
+  value: DTS-SHAREDSERVICES-PROD-Video Hearings
+- group: external
+
+stages:
+
+################################################
+# Check State. #################################
```
Comment (chris-pateman) [inline azure-pipelines.infra-core.yaml:32]: do we need a CI stage first?
https://github.com/hmcts/azure-devops-templates/blob/master/templates/tests/terraform/allTests.yaml

File: azure-pipelines.infra-core.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+- stage: 'Check_Terraform_State'
+  displayName: Check Terraform State
+  jobs:
+  - job: 'Check_Terraform_State'
+    displayName: Check Terraform State
+    steps:
+    - template: pipeline-steps/check-state.yaml
```
Comment (chris-pateman) [inline azure-pipelines.infra-core.yaml:40]: RG and Storage Account should be created per environment and then a state file per env

File: azure-pipelines.infra-core.yaml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+################################################
+# Terraform Plan & Apply. ######################
+
+- ${{ each env in parameters.environment }}:
+  - template: pipeline-steps/terraform-plan-apply.yaml
+    parameters:
+      terraformVersion: ${{ variables.tfVersion }}
+      env: ${{ env }}
+      location: 'UK South'
```
Comment (chris-pateman) [inline azure-pipelines.infra-core.yaml:56]: Can location be a global variable?

File: azure-pipelines.wowza.yaml [hunk 6]
```diff
@@ -0,0 +1,61 @@
+resources:
+  repositories:
+    - repository: templates
+      type: github
+      name: hmcts/azure-devops-templates
+      ref: refs/heads/master
+      endpoint: hmcts
+
+trigger:
```
Comment (chris-pateman) [inline azure-pipelines.wowza.yaml:9]: Same as above?

File: azure-pipelines.wowza.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+   - Dev
+   - Demo
+
+pool:
+  vmImage: 'ubuntu-latest'
+
+variables:
+- name: tfVersion
+  value: 1.1.0
+- name: tfStateServiceConnection
```
Comment (chris-pateman) [inline azure-pipelines.wowza.yaml:25]: same as above

File: azure-pipelines.wowza.yaml [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+################################################
+# Check State. #################################
+
+- stage: 'Check_Terraform_State'
+  displayName: Check Terraform State
+  jobs:
+  - job: 'Check_Terraform_State'
+    displayName: Check Terraform State
+    steps:
+    - template: pipeline-steps/check-state.yaml
```
Comment (chris-pateman) [inline azure-pipelines.wowza.yaml:40]: same as above but share the same RG and Storage Account just different folder

File: pipeline-steps/terraform-apply.yaml [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+      
+  - task: TerraformInstaller@0
+    displayName: 'Install Terraform ${{ parameters.tfversion }}'
+    inputs:
+      terraformVersion: ${{ parameters.tfversion }}
+
+  - task: ExtractFiles@1
+    displayName: 'Extract Terraform Artifacts'
+    inputs:
+      archiveFilePatterns: '$(Pipeline.Workspace)/tfplan-$(Build.BuildId)-${{ parameters.project }}-${{ lower(parameters.environment) }}-${{ parameters.stack }}/tfplan-$(Build.BuildId)-${{ parameters.project }}-${{ lower(parameters.environment) }}-${{ parameters.stack }}.zip'
```
Comment (chris-pateman) [inline pipeline-steps/terraform-apply.yaml:43]: Can this be a variable so its easier to find?

File: pipeline-steps/terraform-plan-apply.yaml [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+      # but it might be better to template the pipeline more and have a different pipeline file
+      # so that someone doesn't pick the wrong var in the dropdown
+      # same comment as above ^^
+    steps:
+    - ${{ if eq(parameters.activityName, 'Video_Hearing_Core') }}:
+      - template: vh-tenant-replacement.yaml
+        parameters:
+          environment: ${{ variables.environment }}
+
+    - ${{ if eq(parameters.activityName, 'Video_Hearing_Core') }}:
```
Comment (chris-pateman) [inline pipeline-steps/terraform-plan-apply.yaml:65]: Could putthis under the same if statement above

File: pipeline-steps/terraform-plan-apply.yaml [hunk 11]
```diff
@@ -0,0 +1,129 @@
+parameters:
```
Comment (chris-pateman) [inline pipeline-steps/terraform-plan-apply.yaml:1]: do we need this if we have a template for plan and a template for appy?

**Review by chris-pateman — CHANGES_REQUESTED** [review]

File: azure-pipelines.infra-core.yaml (hunk 2 above)
Comment (v1ferrarij) [reply azure-pipelines.infra-core.yaml:9]: Yeah, going to add at a later date.

File: azure-pipelines.infra-core.yaml (hunk 3 above)
Comment (v1ferrarij) [reply azure-pipelines.infra-core.yaml:32]: I left this out for a number of reasons:

1.  terraform validate is run in the Plan stage anyway, so seems pointless having it twice.
2. terraform fmt only formats your code to look nice, which is only going to affect the cloned code on the DevOps agent, it isn't going to impact performance of Terraform, the person who writes the code on their local machine should run that anyway after they have completed work and then push to git.
3.  Why are we validating file size?
4. TF Sec was never working (see below)

I don't think we need another stage, if we do want to get TFSEC, or continue to validate file size, I believe we should add this to the plan stage as extra steps.

File: azure-pipelines.infra-core.yaml (hunk 5 above)
Comment (v1ferrarij) [reply azure-pipelines.infra-core.yaml:56]: made location Global Variables.

File: azure-pipelines.wowza.yaml (hunk 8 above)
Comment (v1ferrarij) [reply azure-pipelines.wowza.yaml:40]: see above

File: pipeline-steps/terraform-plan-apply.yaml (hunk 11 above)
Comment (v1ferrarij) [reply pipeline-steps/terraform-plan-apply.yaml:1]: both templates accept the same inputs, so makes sense put them into one, so we are not repeating the code in azure-pipelines.infra-core/wowza.yaml

File: pipeline-steps/terraform-plan-apply.yaml (hunk 10 above)
Comment (v1ferrarij) [reply pipeline-steps/terraform-plan-apply.yaml:65]: amended.

File: azure-pipelines.wowza.yaml (hunk 6 above)
Comment (v1ferrarij) [reply azure-pipelines.wowza.yaml:9]: will amend at later date

File: azure-pipelines.wowza.yaml [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+      parameters:
+        serviceConnection: ${{ variables.tfStateServiceConnection }}
+        resourceGroup: vh-infra-core-tf
+        storageAccountName: vhinfracoretf
+        containerName: tfstate
+        location: 'uk south'
```
[the remark this answers is not in the document]
Comment (v1ferrarij) [reply azure-pipelines.wowza.yaml:46]: set global var

**Review by chris-pateman — CHANGES_REQUESTED** [review]

**Pull request merged**